### PR TITLE
sparse_momentum_op is used to save w@GRAD memory for gather_op

### PR DESCRIPTION
**Pull request by GuoxiaWang**
### PR types

New features

### PR changes

OPs

### Describe

sparse_momentum_op is used to save w@GRAD memory for gather_op when gather from a large parameter, e.g.

```
# for paddle.optimizer.Momentum
weight = paddle.static.create_parameter(
    shape=[512, 10000000],
    dtype='float32',
    name='large_fc_w',
    is_bias=False)
weight = paddle.gather(weight, index, axis=1)
```

### CODE REVIEW

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+  void Make() override;
+};
+
+class SparseMomentumOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+        ctx->HasInput("Param"), true,
+        platform::errors::NotFound(
+            "Input(param) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(ctx->HasInput("Grad"), true,
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+                      platform::errors::NotFound(
+                          "Input(grad) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+        ctx->HasInput("Velocity"), true,
+        platform::errors::NotFound(
+            "Input(velocity) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+        ctx->HasInput("Index"), true,
+        platform::errors::NotFound(
+            "Input(Index) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+        platform::errors::NotFound(
+            "Input(LearningRate) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(ctx->GetInputsVarType("Param").front(),
+                      framework::proto::VarType::LOD_TENSOR,
+                      platform::errors::InvalidArgument(
+                          "The input var's type of SparseMomentum should be "
+                          "LoDTensor, but the received is %s",
+                          ctx->GetInputsVarType("Param").front()));
+
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+        ctx->HasOutput("ParamOut"), true,
+        platform::errors::NotFound(
+            "Output(ParamOut) of SparseMomentum should not be null."));
+    PADDLE_ENFORCE_EQ(
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: Use OP_INOUT_CHECK.

File: paddle/fluid/operators/optimizers/sparse_momentum_op.h [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+                      platform::errors::InvalidArgument(
+                          "Maybe the Input variable LearningRate has not "
+                          "been initialized. You may need to confirm "
+                          "if you put exe.run(startup_program) "
+                          "after optimizer.minimize function."));
+    PADDLE_ENFORCE_EQ(framework::product(lr_dims), 1,
+                      platform::errors::InvalidArgument(
+                          "Learning_rate should be a scalar. But Received "
+                          "LearningRate's dim [%s]",
+                          framework::product(lr_dims)));
```
Comment (sandyhouse) [inline paddle/fluid/operators/optimizers/sparse_momentum_op.h]: The above two statements can be combined into one statement.

**Review by sandyhouse — CHANGES_REQUESTED** [review]

**Review by TCChenlong — DISMISSED** [review]

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM for op_function_generator.cc

**Pull request merged**